---
title: "01_merge_Biop"
output: html_document
date: '2022-02-22'

--- 

<!--
#-------------------------------------------------------------------------------
# This script processes filtered single-cell RNA-seq count matrices stored in
# .h5seurat format, adds metadata information and performs several actions in
# order to perform a first step filtering:
# - Merges them into a single Seurat .h5seurat object for downstream analysis.
# - Removes low quality Samples
# - Calculate Cycle Score
# - Integrate the data using SCT approach
# - Dimensional reduction and clusterization

# Author: Carlos Simon Foundation (Raul PM @supermegamio, Diego AC @diegoamoros)
# Date: 2022-02-22
#-------------------------------------------------------------------------------
# Former  "22022022_AS_merge"
-->

```{r}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
                      error = FALSE, comment = NA, cache = FALSE,
                      R.options = list(width = 320), fig.align = "center",
                      out.width = "95%", out.width = 3160, fig.width = 12,
                      fig.height = 10)
```

```{r}
library(Seurat)
library(SeuratDisk)
library(ggplot2)
library(dittoSeq)
library(ggpubr)
library(yaml)
library(this.path)

## Load config.yml file
find_config <- function(names = c("config.yaml", "config.yml")) {
  currrent.path <- dirname(this.path::this.path())
  dirs <- unique(c(currrent.path, dirname(currrent.path)))
  for (d in dirs) {
    for (nm in names) {
      p <- file.path(d, nm)
      if (file.exists(p))
        return(normalizePath(p, winslash = "/", mustWork = TRUE))
    }
  }
  stop("config.yaml|config.yml not found: ",
       paste(dirs, collapse = " o "), call. = FALSE)
}

config <- yaml.load_file(find_config())



## Count matrices procesed by Cell Bender (h5 format)
input_folder <- config$filtered_data$dir

## QC plots, QC csv and count matrices processed and filtered (h5seurat format)
output_folder <- config$processed_data$dir
dir.create(output_folder, recursive = TRUE, showWarnings = FALSE)

qc_folder <- config$processed_data$qc_dir
dir.create(qc_folder, recursive = TRUE, showWarnings = FALSE)

integration1_folder <- config$processed_data$integration1_dir
dir.create(integration1_folder, recursive = TRUE, showWarnings = FALSE)


## Load metadata
metadata_as <- read.csv(file = config$metadata$file)

files_AS_EpiStroma <- list.files(path = input_folder, full.names = TRUE)
```


Mitocondrial ratio of 0.25 and minimum of 500 genes detected

```{r}

# Remove Sample 11 (there are not post samples)
files_AS_EpiStroma <- files_AS_EpiStroma[which(!(files_AS_EpiStroma %in%
paste0(input_folder,"/",
c("run_count_0192_ASH-21G0001_ENTIRE-1-11-S1_Filter_AS.H5seurat",
"run_count_0192_ASH-21G0002_ENTIRE-1-11-E1_Filter_AS.H5seurat"))))]

# Load samples files
sc_AS_files <- vector(mode = "list", length = length(files_AS_EpiStroma))
for (sc_file in files_AS_EpiStroma) {
 
  nm <- SeuratDisk::LoadH5Seurat(sc_file)
  objname <- gsub(pattern = "_Filter_AS.H5seurat",
                  replacement = "",basename(sc_file))
  objname <- gsub(pattern = "run_count_|run_",
                  replacement = "sc_obj_", objname)
  assign(objname, nm)
}

sc_list <- c(unlist(mget(ls(pattern = "sc_obj"))))
sc_names <- gsub(pattern = ".*\\_",
                 x = ls(pattern = "sc_obj"), replacement = "")
sc_run <- stringr::str_extract(pattern = "[0-9]{4}",
                               string = ls(pattern = "sc_obj"))
sc_run_names <- paste0(sc_run,"_",sc_names)

sc_multiple <- sc_list[[1]]
sc_multiple <- merge(sc_multiple, y = sc_list[2:length(sc_list)],
                     add.cell.ids = sc_names)

rm(list = ls(pattern = "sc_obj.*",))
rm(sc_list)
rm(nm)
gc()
DietSeurat(sc_multiple)
```

## Add metadata

```{r}
Patient <- plyr::mapvalues(
  x = sc_multiple$orig.ident,
  from = metadata_as$Bioinfocode,
  to = metadata_as$Patient)

Cell_type <- plyr::mapvalues(
  x = sc_multiple$orig.ident,
  from = metadata_as$Bioinfocode,
  to = metadata_as$Cell_type)

Treatment_stage <- plyr::mapvalues(
  x = sc_multiple$orig.ident,
  from = metadata_as$Bioinfocode,
  to = metadata_as$Stage_treament)

sc_multiple$Patient <- Patient
sc_multiple$Cell_type <- Cell_type
sc_multiple$Treatment_stage <- Treatment_stage
VlnPlot(object = sc_multiple,
        features = "nFeature_RNA", group.by = "Patient",
        split.by = "Treatment_stage", log = TRUE, split.plot = TRUE)
VlnPlot(object = sc_multiple,
        features = "mitoRatio", group.by = "Patient",
        split.by = "Treatment_stage", log = FALSE, split.plot = TRUE)

```

## FIlter by nFeatures

```{r}

## FIlter by nFeatures 
more_750_gene <- 
  sc_multiple$nFeature_RNA > config$processed_data$nfeature_rna_filt
sc_multiple$more_750_gene <- more_750_gene

VlnPlot(object = sc_multiple, features = "nFeature_RNA",
        group.by = "Treatment_stage",
        split.by = "more_750_gene", log = TRUE, split.plot = TRUE)

table(sc_multiple$Treatment_stage, sc_multiple$more_750_gene)
```


## Cell Cycle Scoring (Phase) calculation

```{r}


s_genes <- c('UBR7','RFC2','RAD51','MCM2','TIPIN','MCM6','UNG','POLD3','WDR76',
             'CLSPN','CDC45','CDC6','MSH2','MCM5','POLA1','MCM4','RAD51AP1',
             'GMNN','RPA2','CASP8AP2','HELLS','E2F8','GINS2','PCNA','NASP',
             'BRIP1','DSCC1','DTL','CDCA7','CENPU','ATAD2','CHAF1B','USP1',
             'SLBP','RRM1','FEN1','RRM2','EXO1','CCNE2','BLM','PRIM1','UHRF1')

g2m_genes <- c('NCAPD2','ANLN','TACC3','HMMR','GTSE1','NDC80','AURKA','TPX2',
               'BIRC5','G2E3','CBX5','RANGAP1','CTCF','CDCA3','TTK','SMC4','
               ECT2','CENPA','CDC20','NEK2','CENPF','HJURP','CKS2','DLGAP5',
               'PIMREG','TOP2A','PSRC1','CDCA8','CKAP2','NUSAP1','KIF23',
               'KIF11','KIF20B','CENPE','GAS2L3','KIF2C','NUF2','ANP32E',
               'LBR','MKI67','CCNB2','CDC25C','HMGB2','CKAP2L','BUB1','CDK1',
               'CKS1B','UBE2C','CKAP5','AURKB','CDCA2','TUBB4B','JPT1')

sc_multiple <- sc_multiple[,which(sc_multiple$more_750_gene)]
VlnPlot(object = sc_multiple,
        features = "nFeature_RNA", group.by = "Patient",
        split.by = "Treatment_stage", log = TRUE, split.plot = TRUE)


sc_multiple <- CellCycleScoring(object = sc_multiple,
                                s.features = s_genes, g2m.features = g2m_genes)
sc_multiple$cc_Phase <- sc_multiple$Phase
sc_multiple$Phase <- sc_multiple$S.Score - sc_multiple$G2M.Score
```


## First Integration


```{r}


sc_multiple <- SCTransform(sc_multiple,
                           vars.to.regress = c("mitoRatio","Phase"),
                           conserve.memory = TRUE)
sc_multiple <- RunPCA(sc_multiple, verbose = FALSE)
sc_multiple <- RunUMAP(sc_multiple, dims = 1:30, verbose = FALSE)

sc_multiple <- FindNeighbors(sc_multiple, dims = 1:30, verbose = FALSE)
sc_multiple <- FindClusters(sc_multiple, verbose = FALSE)
umap1 <- DimPlot(sc_multiple,
                 group.by = "Treatment_stage", label = TRUE,
                 split.by = "Patient", ncol = 3)
umap2 <- FeaturePlot(sc_multiple,
                     features = "nFeature_RNA", cols = c("Blue","Red"),
                     max.cutoff = "q95", min.cutoff = "q5")
umap3 <- FeaturePlot(sc_multiple,
                     features = "KRT7", cols = c("Blue","Red"),
                     min.cutoff = "q5")
ggsave(filename = file.path(integration1_folder,"umap1_patient_group.svg"),
       device = "svg", dpi = 320, width = 14, height = 14, plot = umap1)
ggsave(filename = file.path(integration1_folder,"umap2_Gene_mito.svg"),
       device = "svg", dpi = 320, width = 6, height = 12, plot = umap2 + umap3)



## Save processed Seurat object
SaveH5Seurat(
  object = sc_multiple, 
  filename = file.path(output_folder, config$processed_data$file))
            
```



